parsely/src/main/java/com/parsely/parselyandroid/ParselyTracker.kt

```diff
@@ -0,0 +1,99 @@
+/*
+    Copyright 2016 Parse.ly, Inc.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+package com.parsely.parselyandroid
+
+import android.content.Context
+
+/**
+ * Tracks Parse.ly app views in Android apps
+ *
+ *
+ * Accessed as a singleton. Maintains a queue of pageview events in memory and periodically
+ * flushes the queue to the Parse.ly pixel proxy server.
+ */
+public interface ParselyTracker {
+    public val engagementInterval: Double?
+    public val videoEngagementInterval: Double?
+    public val flushInterval: Long
+
+    public fun engagementIsActive(): Boolean
+    public fun videoIsActive(): Boolean
+
+    public fun trackPageview(
+        url: String,
+        urlRef: String = "",
+        urlMetadata: ParselyMetadata? = null,
+        extraData: Map<String, Any>? = null,
+    )
+
+    public fun startEngagement(
+        url: String,
+        urlRef: String = "",
+        extraData: Map<String, Any>? = null
+    )
+
+    public fun stopEngagement()
+
+    public fun trackPlay(
+        url: String,
+        urlRef: String = "",
+        videoMetadata: ParselyVideoMetadata,
+        extraData: Map<String, Any>? = null,
+    )
+    public fun trackPause()
+    public fun resetVideo()
+    public fun flushEventQueue()
+    public fun flushTimerIsActive(): Boolean
+
+    public companion object {
+        private const val DEFAULT_FLUSH_INTERVAL_SECS = 60
+        private var instance: ParselyTrackerInternal? = null
+
+        /**
+         * Singleton instance accessor. Note: This must be called after [.sharedInstance]
+         *
+         * @return The singleton instance
+         */
+        @JvmStatic
+        public fun sharedInstance(): ParselyTracker? {
+            return instance
+        }
```

Comment:
I am not sure if the additional `sharedInstance` method is helpful or not. I think the following is the intended usage:

* First call `sharedInstance({configuration})`
* Subsequently call `sharedInstance()`

Although this works - the better approach in my opinion is to:

* First call `sharedInstance({configuration})` and assign it to a property (preferably somewhere where you can inject it)
* Then only access the tracker through your local (to the client binary) property and not use the `sharedInstance()`

This approach is better in my opinion, because it doesn't deal with the nullability and somewhat forces the instantiation of the property to be done in a more proper place. It's also how a client would handle the tracker if it wasn't using a singleton pattern.

So, I wonder if we should not give them the option to use the first approach at all. What do you think?